source/slang/parameter-binding.cpp

```diff
@@ -1838,8 +1868,191 @@ void generateParameterBindings(
+    
+    // Try to find rules based on the selected code-generation target
+    auto layoutContext = getInitialLayoutContextForTarget(targetReq);
+
+    // If there was no target, or there are no rules for the target,
+    // then bail out here.
+    if (!layoutContext.rules)
+        return newProgramLayout;
+
+ 
+    // we need to initialize a layout context to mark used registers
+    SharedParameterBindingContext sharedContext;
+    sharedContext.compileRequest = targetReq->compileRequest;
+    sharedContext.defaultLayoutRules = layoutContext.getRulesFamily();
+    sharedContext.programLayout = programLayout;
+
+    // Create a sub-context to collect parameters that get
+    // declared into the global scope
+    ParameterBindingContext context;
+    context.shared = &sharedContext;
+    context.translationUnit = nullptr;
+    context.layoutContext = layoutContext;
+    
+    auto constantBufferRules = context.getRulesFamily()->getConstantBufferRules();
+    structLayout->rules = constantBufferRules;
+
+    UniformLayoutInfo structLayoutInfo;
+    structLayoutInfo.alignment = globalStructLayout->uniformAlignment;
+    structLayoutInfo.size = 0;
+    bool anyUniforms = false;
+    Dictionary<RefPtr<VarLayout>, RefPtr<VarLayout>> varLayoutMapping;
+    for (auto & varLayout : globalStructLayout->fields)
+    {
+        // To recover layout context, we skip generic resources in the first pass
+        // If the var is a generic resource, its resourceInfos will be empty.
+        if (varLayout->resourceInfos.Count() == 0)
+            continue;
+        SLANG_ASSERT(varLayout->resourceInfos.Count() == varLayout->typeLayout->resourceInfos.Count());
+        auto uniformInfo = varLayout->FindResourceInfo(LayoutResourceKind::Uniform);
+        auto tUniformInfo = varLayout->typeLayout->FindResourceInfo(LayoutResourceKind::Uniform);
+        if (uniformInfo)
+        {
+            anyUniforms = true;
+            SLANG_ASSERT(tUniformInfo);
+            structLayoutInfo.size = Math::Max(structLayoutInfo.size, uniformInfo->index + tUniformInfo->count);
+        }
+        for (UInt i = 0; i < varLayout->resourceInfos.Count(); i++)
+        {
+            auto resInfo = varLayout->resourceInfos[i];
+            auto tresInfo = varLayout->typeLayout->resourceInfos[i];
+            SLANG_ASSERT(resInfo.kind == tresInfo.kind);
+            auto usedRangeSet = findUsedRangeSetForSpace(&context, resInfo.space);
+            markSpaceUsed(&context, resInfo.space);
+            usedRangeSet->usedResourceRanges[(int)resInfo.kind].Add(
+                nullptr, // we don't need to track parameter info here
+                resInfo.index,
+                resInfo.index + varLayout->typeLayout->resourceInfos[0].count);
+        }
+        structLayout->fields.Add(varLayout);
+        varLayoutMapping[varLayout] = varLayout;
+    }
+    auto originalGlobalCBufferInfo = programLayout->globalScopeLayout->FindResourceInfo(LayoutResourceKind::ConstantBuffer);
```

Comment:
This won’t work for Vulkan. You need to be checking if `globalScopeLayout` is a `ConstantBufferTypeLayout`, and not for a specific resource kind. Let’s chat about this later. This is fine for now if it passes the tests (which are DX only)

Reply:
There isn't a `ConstantBufferTypeLayout`, do you mean `ParameterGroupLayout`? But even after I check that, I still need to grab existing binding index some where, and I should do that via `ResourceInfo`, right?